// ontimize-web-ngx/ontimize/services/factories.js
import { InjectionToken } from '@angular/core';
import { AppConfig } from '../config/app-config';
import { Util } from '../util/util';
import { OntimizeEEService } from './ontimize-ee.service';
import { OntimizeExportService } from './ontimize-export.service';
import { OntimizeFileService } from './ontimize-file.service';
import { OntimizeService } from './ontimize.service';
import { OntimizeEEPermissionsService, OntimizePermissionsService } from './permissions/o-permissions.module';
import { OTranslateService } from './translate/o-translate.service';
export var O_DATA_SERVICE = new InjectionToken('Ontimize data service');
export var O_TRANSLATE_SERVICE = new InjectionToken('Translate service');
export var O_FILE_SERVICE = new InjectionToken('File uploader service');
export var O_EXPORT_SERVICE = new InjectionToken('Export service');
export var O_PERMISSION_SERVICE = new InjectionToken('Permission service');
export function dataServiceFactory(injector) {
    var serviceClass = getInjectionTokenValue(O_DATA_SERVICE, injector);
    var service = createServiceInstance(serviceClass, injector);
    if (Util.isDefined(service)) {
        return service;
    }
    var config = injector.get(AppConfig).getConfiguration();
    if (typeof (config.serviceType) === 'undefined') {
        return new OntimizeService(injector);
    }
    else if ('OntimizeEE' === config.serviceType) {
        return new OntimizeEEService(injector);
    }
    return createServiceInstance(config.serviceType, injector);
}
export function translateServiceFactory(injector) {
    var serviceClass = getInjectionTokenValue(O_TRANSLATE_SERVICE, injector);
    var service = createServiceInstance(serviceClass, injector);
    return Util.isDefined(service) ? service : new OTranslateService(injector);
}
export function fileServiceFactory(injector) {
    var serviceClass = getInjectionTokenValue(O_FILE_SERVICE, injector);
    var service = createServiceInstance(serviceClass, injector);
    return Util.isDefined(service) ? service : new OntimizeFileService(injector);
}
export function exportServiceFactory(injector) {
    var serviceClass = getInjectionTokenValue(O_EXPORT_SERVICE, injector);
    var service = createServiceInstance(serviceClass, injector);
    if (Util.isDefined(service)) {
        return service;
    }
    var config = injector.get(AppConfig).getConfiguration();
    if (typeof (config.exportServiceType) === 'undefined') {
        return new OntimizeExportService(injector);
    }
    return createServiceInstance(config.exportServiceType, injector);
}
export function permissionsServiceFactory(injector) {
    var serviceClass = getInjectionTokenValue(O_PERMISSION_SERVICE, injector);
    var service = createServiceInstance(serviceClass, injector);
    if (Util.isDefined(service)) {
        return service;
    }
    var config = injector.get(AppConfig).getConfiguration();
    if (!Util.isDefined(config.permissionsServiceType) || 'OntimizePermissions' === config.permissionsServiceType) {
        return new OntimizePermissionsService(injector);
    }
    else if ('OntimizeEEPermissions' === config.permissionsServiceType) {
        return new OntimizeEEPermissionsService(injector);
    }
    return createServiceInstance(config.permissionsServiceType, injector);
}
function getInjectionTokenValue(token, injector) {
    var service;
    try {
        service = injector.get(token);
    }
    catch (e) {
    }
    return service;
}
function createServiceInstance(clazz, injector) {
    if (!Util.isDefined(clazz)) {
        return;
    }
    var newInstance = Object.create(clazz.prototype);
    clazz.apply(newInstance, [injector]);
    return newInstance;
}
