// ontimize-web-ngx/ontimize/components/table/extensions/dialog/export/o-table-export-dialog.component.js
import { HttpErrorResponse } from '@angular/common/http';
import { ChangeDetectionStrategy, Component, Inject, Injector, ViewEncapsulation } from '@angular/core';
import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material';
import { Subscription } from 'rxjs';
import { filter } from 'rxjs/operators';
import { DialogService, OntimizeExportService, OTranslateService } from '../../../../../services';
import { exportServiceFactory } from '../../../../../services/factories';
import { Codes, SQLTypes, Util } from '../../../../../utils';
import { OTableExportButtonService } from '../../export-button/o-table-export-button.service';
var OTableExportConfiguration = (function () {
    function OTableExportConfiguration() {
    }
    return OTableExportConfiguration;
}());
export { OTableExportConfiguration };
var ɵ0 = exportServiceFactory;
var OTableExportDialogComponent = (function () {
    function OTableExportDialogComponent(dialogRef, injector, config) {
        this.dialogRef = dialogRef;
        this.injector = injector;
        this.config = config;
        this.subscription = new Subscription();
        this.dialogService = injector.get(DialogService);
        this.translateService = this.injector.get(OTranslateService);
        this.oTableExportButtonService = this.injector.get(OTableExportButtonService);
        if (config && Util.isDefined(config.visibleButtons)) {
            this.visibleButtons = Util.parseArray(config.visibleButtons.toLowerCase(), true);
        }
    }
    OTableExportDialogComponent.prototype.ngOnInit = function () {
        this.initialize();
    };
    OTableExportDialogComponent.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
    };
    OTableExportDialogComponent.prototype.initialize = function () {
        var _this = this;
        this.configureService();
        this.subscription.add(this.oTableExportButtonService.export$.pipe(filter(function (type) { return ['xlsx', 'html', 'pdf'].indexOf(type) === -1; })).subscribe(function (e) { return _this.export(e); }));
    };
    OTableExportDialogComponent.prototype.configureService = function () {
        var loadingService = OntimizeExportService;
        if (this.config.serviceType) {
            loadingService = this.config.serviceType;
        }
        this.exportService = this.injector.get(loadingService);
        var serviceCfg = this.exportService.getDefaultServiceConfiguration(this.config.service);
        this.exportService.configureService(serviceCfg, Codes.EXPORT_MODE_ALL === this.config.mode);
    };
    OTableExportDialogComponent.prototype.export = function (exportType, button) {
        if (button) {
            button.disabled = true;
        }
        var exportData = {
            data: this.config.data,
            columns: this.config.columns,
            columnNames: this.config.columnNames,
            sqlTypes: this.config.sqlTypes,
            filter: this.config.filter
        };
        var self = this;
        this.proccessExportData(exportData.data, exportData.sqlTypes);
        this.exportService.exportData(exportData, exportType, this.config.entity)
            .subscribe(function (res) { return self.dialogRef.close(true); }, function (err) { return self.handleError(err); });
    };
    OTableExportDialogComponent.prototype.proccessExportData = function (data, sqlTypes) {
        var _this = this;
        Object.keys(sqlTypes).forEach(function (key) {
            if (SQLTypes.BOOLEAN === sqlTypes[key]) {
                var yes_1 = _this.translateService.get('YES');
                var no_1 = _this.translateService.get('NO');
                data.forEach(function (row) {
                    if (row[key]) {
                        row[key] = Util.parseBoolean(row[key]) ? yes_1 : no_1;
                    }
                });
            }
        });
    };
    OTableExportDialogComponent.prototype.isButtonVisible = function (btn) {
        return !this.visibleButtons || (this.visibleButtons.indexOf(btn) !== -1);
    };
    OTableExportDialogComponent.prototype.handleError = function (err) {
        console.error(err);
        var self = this;
        if (err instanceof HttpErrorResponse) {
            this.dialogService.alert('ERROR', err.message).then(function () { return self.dialogRef.close(false); });
        }
        else {
            this.dialogService.alert('ERROR', 'MESSAGES.ERROR_EXPORT_TABLE_DATA').then(function () { return self.dialogRef.close(false); });
        }
    };
    OTableExportDialogComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-table-export-dialog',
                    template: "\n    <span mat-dialog-title>{{ 'TABLE.BUTTONS.EXPORT' | oTranslate }}</span>\n    <mat-dialog-content>\n      <div mat-subheader>{{ 'TABLE.DIALOG.EXPORT.DESCRIPTION' | oTranslate }}</div>\n      <div fxLayout=\"row wrap\" fxLayoutAlign=\"space-around center\" fxLayoutGap=\"8px\">\n        <o-table-export-button #excelButton *ngIf=\"isButtonVisible('excel')\" svg-icon=\"ontimize:EXCEL\" label=\"TABLE.BUTTONS.EXCEL\" export-type=\"xlsx\"\n          (onClick)=\"export('xlsx', excelButton)\" class=\"excel-button\"></o-table-export-button>\n        <o-table-export-button #htmlButton *ngIf=\"isButtonVisible('html')\" svg-icon=\"ontimize:HTML\" label=\"TABLE.BUTTONS.HTML\" export-type=\"html\"\n          (onClick)=\"export('html', htmlButton)\" class=\"html-button\"></o-table-export-button>\n        <o-table-export-button #pdfButton *ngIf=\"isButtonVisible('pdf')\" svg-icon=\"ontimize:PDF\" label=\"TABLE.BUTTONS.PDF\" export-type=\"pdf\"\n          (onClick)=\"export('pdf', pdfButton)\" class=\"pdf-button\"></o-table-export-button>\n        <ng-container *ngTemplateOutlet=\"config.options\"></ng-container>\n      </div>\n    </mat-dialog-content>\n\n    <mat-dialog-actions fxLayoutAlign=\"end center\">\n      <button type=\"button\" mat-stroked-button [mat-dialog-close]=\"false\">{{ 'CANCEL' | oTranslate | uppercase }}</button>\n    </mat-dialog-actions>\n  ",
                    styles: ["\n    .o-table-export-dialog .mat-icon{padding:6px 6px 0;width:48px;height:48px;font-size:48px}\n  "],
                    providers: [
                        { provide: OntimizeExportService, useFactory: ɵ0, deps: [Injector] }
                    ],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        'class': 'o-table-export-dialog'
                    },
                    encapsulation: ViewEncapsulation.None
                },] },
    ];
    OTableExportDialogComponent.ctorParameters = function () { return [
        { type: MatDialogRef },
        { type: Injector },
        { type: OTableExportConfiguration, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] }
    ]; };
    return OTableExportDialogComponent;
}());
export { OTableExportDialogComponent };
export { ɵ0 };
