// ontimize-web-ngx/ontimize/components/table/extensions/export-button/o-table-export-button.component.js
import { Component, EventEmitter, Injector } from '@angular/core';
import { OTableExportButtonService } from './o-table-export-button.service';
export var DEFAULT_INPUTS_O_TABLE_EXPORT_BUTTON = [
    'icon',
    'svgIcon : svg-icon',
    'olabel: label',
    'exportType: export-type'
];
export var DEFAULT_OUTPUTS_O_TABLE_EXPORT_BUTTON = [
    'onClick'
];
var OTableExportButton = (function () {
    function OTableExportButton(injector) {
        this.injector = injector;
        this.onClick = new EventEmitter();
        this.oTableExportButtonService = this.injector.get(OTableExportButtonService);
    }
    OTableExportButton.prototype.click = function () {
        this.onClick.emit(this.exportType);
        this.oTableExportButtonService.export$.next(this.exportType);
    };
    OTableExportButton.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-table-export-button',
                    template: "\n    <button type=\"button\" mat-raised-button (click)=\"click()\">\n      <div fxLayout=\"column\" fxLayoutAlign=\"center center\">\n        <mat-icon class=\"material-icons o-button-icon\" *ngIf=\"icon!==undefined\" style=\"vertical-align:middle\">{{ icon }}</mat-icon>\n        <mat-icon class=\"material-icons o-button-icon\" *ngIf=\"svgIcon!==undefined\" style=\"vertical-align:middle\" [svgIcon]=\"svgIcon\"></mat-icon>\n        <span>{{ olabel | oTranslate }}</span>\n      </div>\n    </button>\n  ",
                    inputs: DEFAULT_INPUTS_O_TABLE_EXPORT_BUTTON,
                    outputs: DEFAULT_OUTPUTS_O_TABLE_EXPORT_BUTTON
                },] },
    ];
    OTableExportButton.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    return OTableExportButton;
}());
export { OTableExportButton };
