// ontimize-web-ngx/ontimize/components/form/guards/o-form-can-deactivate.guard.js
import { Injectable, Injector } from '@angular/core';
import { Codes } from '../../../utils';
import { LoginService } from '../../../services/login.service';
var CanDeactivateFormGuard = (function () {
    function CanDeactivateFormGuard(injector) {
        this.injector = injector;
        this.loginService = this.injector.get(LoginService);
    }
    CanDeactivateFormGuard.prototype.canDeactivate = function (component, curr, state, future) {
        if (!this.loginService.isLoggedIn()) {
            return true;
        }
        var futureQueryParams = future.root.queryParams;
        if (futureQueryParams.hasOwnProperty(Codes.IGNORE_CAN_DEACTIVATE)) {
            return true;
        }
        if (this.oForm) {
            return this.oForm.canDeactivate();
        }
        return true;
    };
    CanDeactivateFormGuard.prototype.setForm = function (form) {
        this.oForm = form;
    };
    CanDeactivateFormGuard.CLASSNAME = 'CanDeactivateFormGuard';
    CanDeactivateFormGuard.decorators = [
        { type: Injectable },
    ];
    CanDeactivateFormGuard.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    return CanDeactivateFormGuard;
}());
export { CanDeactivateFormGuard };
